ext/cuda/bilinearinterpolation.jl

```diff
@@ -0,0 +1,222 @@
+
+Interpolate `fsource`, defined on source grid, onto the target grid.
+The horizontal source and target grids are defined in `bilinear`.
+Here `fsource` is an N-dimensional array where the last two dimensions are assumed to be horizontal dimensions.
+
+For example, `fsource` can be of size `[n1, n2..., nx, ny]`, where `nx` and `ny` are the horizontal dimensions.
+Single horizontal level is also supported.
+The number of horizontal levels should be same for both source and target arrays.
+
+The interpolation is parallelized by distributing the horizontal target points across a `(nxblocks, nyblocks)` sized CUDA grid.
+All vertical levels are placed on the same CUDA block.
+When `nlevels` exceeds `nthreads`, `nzloops` loops are executed by each thread to process all vertical levels.
+Similarly, when `ntargetx` exceeds `nxblocks` or `ntargety` exceeds `nyblocks`, `nxloops`/`nyloops` are executed
+by all threads in a given block per that direction.
+"""
+function interpolatebilinear!(
+    ftarget::CuArray{FT, N},
+    bilinear::B,
+    fsource::CuArray{FT, N},
+) where {FT, N, B}
+    @assert N ≥ 2
+    (leveldimssource..., nxs, nys) = size(fsource)
+    (leveldimstarget..., nxt, nyt) = size(ftarget)
+
+    nsourcex, nsourcey, ntargetx, ntargety = get_dims(bilinear)
+
+    # ensure number of vertical levels are same between fsource and ftarget
+    @assert leveldimssource == leveldimstarget
+    # ensure horizontal dimensions of fsource match with horizontal source grid dimensions 
+    # in bilinear 
+    @assert nxs == nsourcex && nys == nsourcey
+    # ensure horizontal dimensions of ftarget match with horizontal target grid dimensions
+    # in bilinear
+    @assert nxt == ntargetx && nyt == ntargety
+    # max threadgroups per grid is not officially specified by Metal
+    # It is possible this number is unlimited on newer Apple models (>m3)
+    maxxblocks = 1024
+    maxyblocks = 1024
+    # target n loops per dimension per block/threadgroup
+    targetloopsperblockdimx = 4
+    targetloopsperblockdimy = 4
+    nxblocks = min(cld(ntargetx, targetloopsperblockdimx), maxxblocks)
+    nyblocks = min(cld(ntargety, targetloopsperblockdimy), maxyblocks)
+
+    nxloops = cld(ntargetx, nxblocks)
+    nyloops = cld(ntargety, nyblocks)
+
+    levelcidxs = CartesianIndices(leveldimssource)
+    nlevels = length(levelcidxs)
+
+    kernel = @cuda launch = false interpolatebilinear_kernel!(
+        ftarget,
+        bilinear,
+        fsource,
+        levelcidxs,
+        (nxloops, ntargetx),
+        (nyloops, ntargety),
+        (1, 1),
+    )
+    kernel_config = CUDA.launch_configuration(kernel.fun)
+    nthreads = min(kernel_config.threads, nlevels)
+    nzloops = cld(nlevels, nthreads)
+
+    @cuda threads = nthreads blocks = (nxblocks, nyblocks) interpolatebilinear_kernel!(
+        ftarget,
+        bilinear,
+        fsource,
+        levelcidxs,
+        (nxloops, ntargetx),
+        (nyloops, ntargety),
+        (nzloops, nlevels),
+    )
```

Comment:
This seems odd to me: why are we determining a launch configuration and then passing different arguments when launching? Why not use `kernel()` here? I'm concerned that `launch_configuration` may not always give the correct configuration for when this is called.

Let's please fix this.

Reply:
`nzloops` cannot be determined ahead of time before `max_threads` is available. Each column is placed on a different block. Hence the difference in the last argument between the two calls.

Reply:
I'm concerned that this is still not safe, and we've had issues with this in the past. For example, using `@inbounds` can change the number of threads allocated for a given kernel.

This reminds me of how Tim's explanation of how to handle this sort of dependency with dynamic shared memory. But here, we're not using dynamic shared memory: https://www.youtube.com/watch?v=plgbsH9f7gI (at 2:52:30)

Reply:
Yes. We are not using any shared memory here. Since the intent is to loop over `zthreads`, I believe resource estimate for a loop size of 1 is a good estimate.  I am not that worried about safety in this specific case. Another option is to use a conservative size of `256` that is guaranteed to work!